### ISSUE TITLE
sftp put naming incorrect with str input of windows path

### ISSUE THREAD

**Issue by NewUserHa** [ronf/asyncssh#219]
`await sftp.put('D:\###B\#R\s.m')`
will result:
`/root/"D:\###B\#R\s.m"`

**Comment by ronf** [comment]
Are you copying from a Windows to a UNIX system here? If so, this is expected behavior. Since you only provided a single argument, it's going to pass that filename in the SFTP request as-is on the wire, and it is up to the remote system to decide how to interpret it. If you use a Windows-style path in your call, this may work correctly when copying to another Windows system, but it's not likely to do what you want when going between Windows and UNIX. You'll need to pass in a second argument (remotepath) to specify a pathname which matches the path syntax expected by the remote system.

**Comment by NewUserHa** [comment]
```
import asyncio
import asyncssh

async def a():
    conn = await asyncssh.connect()
    async with conn.start_sftp_client() as sftp:
        await sftp.put(r'D:\###B\#R\sm', '/root')

asyncio.run(a())
```
result the same:
```
/root/D:\###B\#R\s.m
```

**Comment by ronf** [comment]
Yeah - if you are copying multiples files into a directory, there's a similar issue where the source path must using POSIX-style path separators (forward slash instead of backslash). That should work even on Windows, and may solve your problem here. In fact, looking more closely at the code, I think it may solve things in the single-argument case as well.

I can look into automatically converting backslashes to forward slashes when on Windows, and perhaps automatically stripping off the drive letter as well. However, as a workaround, I think you should be ok as long as there's a slash in the source path which separates the final portion in the path that you want to keep on the target system from everything else. Nothing else from the source path will be used in that case.

**Comment by NewUserHa** [comment]
ok. thanks.

**Comment by ronf** [comment]
I got a chance to look at this today, and I've made a change in commit 720e23a in the "develop" branch that I think will address the problem you were seeing here.

AsyncSSH will now use the local OS's basename function when extracting a filename from a local path that is used to construct remote paths. So, regardless of whether forward or back slashes are used, it should now work correctly. It also properly strips off drive letters in a local path when doing this.

If you get a chance, please let me know if this fixes the problems you saw. It should now work with or without a destination path being specified, and also work with a destination path that is a directory.

**Comment by NewUserHa** [comment]
tested. ok now.

**Comment by ronf** [comment]
Great, thanks!

**Comment by NewUserHa** [comment]
just noticed, there's a log `WARNING:asyncssh:Unable to read keys from agent: [Errno 2] No such file or directory: '\\\\.\\pipe\\openssh-ssh-agent'`
I don't feel it exist before but it does exist even after I uninstall and install asyncssh from pip.
is it normal?

**Comment by ronf** [comment]
Yeah - there was a bug fixed in AsyncSSH 1.17.0 related to not automatically attempting to use the SSH agent on Windows. See #217 for more details. A consequence of that fix is that it will produce this extra log message if you don't have either Pageant or the Windows 10 OpenSSH agent running. You can set 'agent_path=None' explicitly to get rid of this warning.

When I looked at this before, I didn't think this would be easy to fix, but tonight I took another look and think I might be able to do something to not print this warning if the user hasn't explicitly set an agent path. As part of this, I think I can also do some other cleanup of this code.

If all goes well, I'll try and get that in before I make the next release.

**Comment by ronf** [comment]
I have a fix for this -- it is now in the "develop" branch as commit 8749f84, and I'll be rolling it into a new release just as soon as I confirm the new build looks good.

**Comment by NewUserHa** [comment]
it seems disappeared now.

**Comment by ronf** [comment]
Great - thanks for the confirmation!

**Comment by NewUserHa** [comment]
ehh...
I accidentally found your site about your interests.
why do you collect anime DVDs? it's usually to collect BD and many animes has BD reversion and those are usually better on visually.

**Comment by ronf** [comment]
This is a bit off-topic for AsyncSSH. :)

Since you asked, though, pretty much all my anime DVDs were bought before Blu-Ray existed. Most of my collection dates back 15-20 years at this point on DVD and I had a pretty good sized Laserdisc collection before that. I do have a handful of Blu-Ray discs (mostly sci-fi movies), but most of what I buy these days is through one of the online streaming services.

**Comment by NewUserHa** [comment]
I can't open a new issue for this.

I suggest BD instead of streaming since visual quality.

**Comment by ronf** [comment]
Agreed about Blu-Ray quality, and that's what I would look for if I were still buying physical media. I just don't do that very often these days.

**Comment by NewUserHa** [comment]
`asyncssh.scp(r'D:/abc', 'addr:/', ...)` will cause 
```
Traceback (most recent call last):
...
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\scp.py", line 882, in scp
    **kwargs)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\scp.py", line 82, in _parse_path
    conn = yield from connect(conn, **kwargs)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\misc.py", line 185, in __iter__
    return (yield from self._coro)
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 5751, in connect
    'Opening SSH connection to'))
  File "C:\Program Files\Python37\lib\site-packages\asyncssh\connection.py", line 165, in _connect
    local_addr=local_addr)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 910, in create_connection
    type=socket.SOCK_STREAM, proto=proto, flags=flags, loop=self)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 1279, in _ensure_resolved
    proto=proto, flags=flags)
  File "C:\Program Files\Python37\lib\asyncio\base_events.py", line 789, in getaddrinfo
    None, getaddr_func, host, port, family, type, proto, flags)
  File "C:\Program Files\Python37\lib\concurrent\futures\thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "C:\Program Files\Python37\lib\socket.py", line 748, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 11001] getaddrinfo failed
```

because `D:` and `def _parse_path(path, **kwargs):` in scp.py of asyncssh:
```
def _parse_path(path, **kwargs):
    """Convert an SCP path into an SSHClientConnection and path"""

    from . import connect

    if isinstance(path, tuple):
        conn, path = path
    elif isinstance(path, str) and ':' in path: # this line
        conn, path = path.split(':')
    elif isinstance(path, bytes) and b':' in path:
        conn, path = path.split(b':')
    elif isinstance(path, (str, bytes, PurePath)):
        conn = None
    else:
        conn = path
        path = b'.'

    if isinstance(conn, (str, bytes)):
        close_conn = True
        conn = yield from connect(conn, **kwargs)
    elif isinstance(conn, tuple):
        close_conn = True
        conn = yield from connect(*conn, **kwargs)
    else:
        close_conn = False

    return conn, path, close_conn
```

**Comment by NewUserHa** [comment]
and scp still has the `/root/D:\###B\#R\s.m` issue.

**Comment by ronf** [comment]
Ah, yes - the 'D:' is ambiguous there. It could refer to a host named "D" or it could refer to a drive letter of a path on the local host. Right now, AsyncSSH assumes anything with a ':' refers to a remote host. You should be able to work around this for now by replacing 'D:/abc' with a tuple of the form (None, 'D:/abc'), but I could potentially change this to treat single-letter values before the colon on Windows as local paths by default, and require you write ('d', '/abc') if you really had a host named "d".

Regarding the issue with backslashes, I'll take a look tonight.

**Comment by NewUserHa** [comment]
hm. I tried to modify to make it work. found single-letter before seems to be a solution.
but it's hard to say is it "on Windows" of a path before detecting if starts with single-letter.

but if there's only one colon in it and starts with a single letter, it's without conn. Elif two colon, split(':',1). that may be a solution not to make windows feel sad since using ('d', '/abc')

**Comment by ronf** [comment]
Thanks for the report! In addition to three issues you identified here, I also found another small issue with pathname normalization when the "chroot" argument is used. These should all be fixed in commit 8c2a367, now checked into the "develop" branch:

* As you suggested, any SCP argument that has a single letter followed by colon is now treated as a drive letter rather than a remote hostname, but only when running on Windows (tested using sys.platform). If you really want to connect to a single-letter remote host, the tuple syntax can be used instead.
* When multiple colons are present, I made the change you suggested about adding the ", 1" in the call to split(), so that any colons after the first just become part of the path instead of producing a tuple with the wrong number of members.
* When parsing the base filename out of a path, the OS "basename" is now used in SCP, similar to the change I previously made for SFTP, allowing the drive letter and path to be stripped off correctly when Windows-style pathnames are passed in.
* I fixed a small error in constructing the path name when the SFTPServer "chroot" feature was used and the client passed in a filename that looked like it contained a drive letter.
